### PR TITLE
fix(MeshHTTPRoute): isolate listeners

### PR DESCRIPTION
**Pull request by michaelbeaumont**
The idea behind this PR is to make sure the effects of MeshHTTPRoutes attached to MeshGateway listeners are limited to those listeners, thus the need to map `(address, port, hostname)` to routes instead of `(address, port)`.

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/02.policies.yaml [hunk 1]
```diff
@@ -3,68 +3,96 @@ type: MeshAccessLog
...
+    - port: 8082
+      protocol: HTTP
+      tags:
+        listener: three
+    - port: 8083
+      protocol: HTTP
+      hostname: four-hostname-1
+      tags:
+        listener: four-hostname-1
+    - port: 8083
```
Comment (bartsmykla) [inline pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/02.policies.yaml:94]: Is it intentional to have the same port as a listener above?

File: pkg/plugins/runtime/gateway/match/hostnames.go [hunk 2]
```diff
@@ -35,6 +35,9 @@ func (h *hostname) contains(n hostname) bool {
 }
 
 func makeHostname(name string) hostname {
+	if name == "" {
+		name = "*"
+	}
```
Comment (slonka) [inline pkg/plugins/runtime/gateway/match/hostnames.go:40]: we have a bunch of these scattered around the codebase, should we make them point to this function?

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml [hunk 3]
```diff
@@ -39,67 +39,133 @@ FromRules:
...
+        modificationTime: "0001-01-01T00:00:00Z"
+        name: gateway
+        type: MeshAccessLog
+      - creationTime: "0001-01-01T00:00:00Z"
+        mesh: mesh-1
+        modificationTime: "0001-01-01T00:00:00Z"
+        name: mesh
+        type: MeshAccessLog
+      Subset: []
+  ByListenerAndHostname:
```
Comment (slonka) [inline pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml:107]: Isn't just `ByListener` a `ByListenerAndHostname` with `*` as a hostname? Does having two entries give us anything else?

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml [hunk 4]
```diff
@@ -39,67 +39,133 @@ FromRules:
...
+    127.0.0.1:8080:*:
```
Comment (slonka) [inline pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml:108]: Isn't this notation problematic for ipv6? should we use a different delimiter?

**Review by slonka — COMMENTED** [review]
I understand the idea and agree with it, left a couple of Qs.

**Review by Automaat — APPROVED** [review]

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml (hunk 3 above)
Comment (michaelbeaumont) [reply pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml:107]: In some sense yes, but the key thing is that it depends on the policy whether targeting hostnames are supported and currently only MeshHTTPRoute supports it. So for MeshHTTPRoute only `ByListenerAndHostname` is set and for other policies only `ByListener`. It makes the distinction a little clearer that for other policies the hostname can _only_ be `*`.

File: pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml (hunk 4 above)
Comment (michaelbeaumont) [reply pkg/plugins/policies/core/matchers/testdata/matchedpolicies/meshgateways/01.golden.yaml:108]: We only serialize this for tests

**Review by slonka — APPROVED** [review]

File: pkg/plugins/runtime/gateway/match/hostnames.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/gateway/match/hostnames.go]: tried to reduce the `"" => "*"` logic in the codebase

**Pull request merged**